### ISSUE TITLE
Recurring Jobs not Running Correctly when summer time ends in BST

### ISSUE THREAD

**Issue by DownChapel** [HangfireIO/Hangfire#567]
After the recent bug with time zones when moving into DST (https://github.com/HangfireIO/Hangfire/issues/529), I have been testing Hangfire v1.5.6 moving out of DST.

I am testing in British Standard Time, where on the last Sunday of October, the time at 2am changes back to 1am. When I test this in Hangfire, my jobs that run every minute, do not run for an hour when the clocks go back from 2am to 1am. However, they do start up again after that hour.

I have narrowed it down to the code in `ScheduleInstant`: https://github.com/HangfireIO/Hangfire/blob/v1.5.6/src/Hangfire.Core/Server/ScheduleInstant.cs#L46-L55

The following test code using 1:59 in BST (0:59 in UTC) should return 1:00 in BST (1:00 UTC). However, it returns 2:00 in BST (2:00 UTC). This is because `NCrontab` doesn't deal with time zones, so if you pass 1:59. you will get 2:00am back.

```
var gmtStandardTime = "GMT Standard Time";
var gmtStandardTimeZone = TimeZoneInfo.FindSystemTimeZoneById(gmtStandardTime);

var crontabSchedule = CrontabSchedule.Parse("* * * * *");

// 0:59 in UTC = 1:59 in BST. 1:00 UTC = 1:00 BST

var utcStartTime = new DateTime(2016, 10, 30, 0, 59, 0);
var startTimeInBst = TimeZoneInfo.ConvertTimeFromUtc(utcStartTime, gmtStandardTimeZone);
var endTimeInBst = TimeZoneInfo.ConvertTimeFromUtc(new DateTime(2016, 10, 30, 2, 20, 0), gmtStandardTimeZone);

var list = crontabSchedule
    .GetNextOccurrences(
        startTimeInBst,
        endTimeInBst)
    .ToList();
```

I'm happy to try and fix this, but I'm not sure how to go about it. Any tips would be appreciated!

**Comment by odinserj** [comment]
@DownChapel, thanks for reporting this! Indeed, NCrontab doesn't handle daylight savings at all, and works correctly only for UTC timezone as it uses simple second/minute/whatever increment in its [`GetNextOccurrence`](https://github.com/atifaziz/NCrontab/blob/master/NCrontab/CrontabSchedule.cs#L212) method, without the knowledge about timezones and daylight savings. As a result – we have missed values when the clock is moved backwards.

The solution proposed in #568 is not enough, because the fixed property is only used for calculating the *estimated time* for the next run. The other method (`GetNextInstants`) is used to decide to run or not to run the recurring job, and it is still prone to errors.

To fix this, we need to teach the `CronSchedule` take the timezones into account when iterating over the time, making it resilient to daylight savings in a complete way as in [Paul Vixie cron](http://linux.die.net/man/8/cron):

> Local time changes of less than three hours, such as  those  caused  by  the  start or end of Daylight Saving Time, are handled specially.  This only applies to jobs that run at a specific time and jobs that are  run with  a    granularity  greater  than  one hour.  Jobs that run more frequently are scheduled normally.
> If time has moved forward, those jobs that would have run in the interval that has been skipped will be run immediately.  Conversely, if time has moved backward, care is taken to avoid running jobs twice.
> Time changes of more than 3 hours are considered to be  corrections  to the clock or timezone, and the new time is used immediately.

Another implementation of CRON format parser is near (please see 494), maybe it's worth to implement the fix there (currently it lacks of support daylight savings too).

**Comment by jcoutch** [comment]
@odinserj - I spent the evening researching time zone support in .NET...and came to the conclusion that it's going to take a while to think up a true solution.

With NCrontab (as well as NCrontab.Advanced), it's a date parsing library that doesn't rely on system time like a true Cron server would, wouldn't be able to detect the moment DST changes occur, and has no direct history of previous runs...implementing something similar to the Paul Vixie cron algorithm would be difficult.  Also, I found out the System.DateTime structure doesn't care about DST, so 1:30 before DST ends = 1:30 after DST ends as far as the .NET Framework is concerned.

I could add a setting in NCrontab.Advanced that when enabled, would use TimeZoneInfo.IsAmbiguousTime against the system's TimeZoneInfo (or one passed in) to see if the time can be represented by two different UTC times, and re-checking that hour for matches.  This would handle runs within that hour, but in instances where the schedule is "30 * * * *", depending on how Hangfire handles it, it could lead to an instance running at 1:30 before DST with GetNextOccurence returning 1:30, 1:30 after DST with GetNextOccurence returning 1:30, and then running at 2:00 (since it would think that the occurence at 1:30 was missed), then again at 2:30.  If that could be worked around in Hangfire, this is a fairly easy change.

Ideally, as suggested in #568, UTC time should be used, NCrontab.Advanced would pick a next occurence based on a passed in TimeZoneInfo object (or system default time zone), and return the next occurence in UTC.  But, this would require a near-complete re-write of the next occurence algorithm, and I haven't thought of a good way to do that while maintaining the performance benefits and taking into account offsets that are not on hour boundaries...like Australia - Cocos Islands (UTC+6:30) and New Zealand (UTC+12:45).

**Comment by vitorpereirafarfetchcom** [comment]
Regarding System.DateTime use NodaTime as a wrapper.
An overview of why System.DateTime is a problem http://blog.nodatime.org/2011/08/what-wrong-with-datetime-anyway.html

Also, a time sensitive application like HangFire shouldn't use System.DateTime directly on its classes but should instead receive a date time provider (on ctor) to allow better Unit Testing.

Regarding time-zone chaos:
* UTC doesn't fix everything

    If the application is sensitive to seconds (HangFire doesn't appear to be) leap seconds can also cause chaos. PS: Check http://infiniteundo.com/post/25326999628/falsehoods-programmers-believe-about-time
* The root problem is NCrontab design

    Cron expressions don't have enough verbosity to deal with DST and their design also lead to a lot of ambiguity on DST. A Cron-expresison to run every hour when DST happens (-1 h) should run twice (and hour has passed) or should run once (the job usually just runs 24 times a day and the 1AM has already happened once)? Also, if I schedule to run once everyday on 1AM should run twice on DST or just once?
    
    On the short-term, sure, something has to be work around, but on the long run, cron expressions should be dumped.

**Comment by DownChapel** [comment]
I like the cron format for how easy it is to declare the schedule. However, I’m not sure of the alternatives. Any solution that allows you to specify a schedule (not in UTC) is going to have to deal with time zones changing.

Although not ideal, I can cope with the issues of multiple/no occurrences of a schedule during a time zone change.  

I agree `NodaTime` makes problems with time zones much more explicit (e.g. throwing an exception if you try and convert to an ambiguous time). The calculation of the next occurrence needs to take the time zone into account. It does look like that would be quite a large change to the NCrontab.Advanced library though!

**Comment by jcoutch** [comment]
Sorry for the delay in responding back.  I agree, NodaTime would be a good addition to NCrontab.Advanced, and would require some significant changes.  But, since Hangfire doesn't use NodaTime, there could still be discrepancies between the conversion of `ZonedDateTime` to `DateTime`.

Maybe the solution is to set the system time to UTC and add a time-zone component in NCrontab.Advanced...perhaps `* * * * * TZ=BST` (analogous to the CRON_TZ variable.)  Incrementing to the next instance would be done in UTC, converted to the specified time zone, then verified against the next match.  Hangfire could store job instances using UTC.  You'd get the benefit of jobs continuing to run after time "falls back", and it'd properly skip over the "spring forward" hour.

For servers that are required to be in a certain time zone, time could be stored in UTC in the system clock, and the `RealTimeIsUniversal` registry entry could be used.  Hangfire would then use the UTC time instead of the timezone the Windows instance is currently in.  For *nix-based systems using Mono, those systems can also be set to use UTC in the system clock.

Thoughts?

**Comment by mgmccarthy** [comment]
@jcoutch, I'm just joining this discussion now, b/c [the project I'm working on](https://github.com/HTBox/allReady) intends to use Hangfire for scheduling jobs.

We are trying to pick our approach to handling timezones and DST. Two things we're knocking around:
- store everything as UTC and always have a time zone component. This way, we can always convert from UTC back to "local" and vice-versa when reading/writing data.
- store everything as DateTimeOffset. The advantage here is you're storing the offset along with a DateTime, but it COULD be problematic if time zone rules change (which they do) and we've already stored a DateTimeOffset for a future entity in our db. In this case, the offset would be wrong.

I don't know anything about NCrontab.Advanced, but your comment above:
"Maybe the solution is to set the system time to UTC and add a time-zone component in NCrontab.Advanced...perhaps * * * * * TZ=BST (analogous to the CRON_TZ variable.)"

if right on. Use/store DateTime as UTC and add a TimeZone component. This way, you'll always be able to convert.

**Comment by MarLoe** [comment]
Might be out of context, but still:

I experienced that winding back the system clock would stop recurrent jobs from running, until the time has catch'ed up.

My experience was as follows:
I started my project on a server, adding some recurrent jobs. When they ran, I realized that the output of these jobs had the wrong date. I had set the wrong month. So i turned back the clock one month, and then recurrent jobs stopped executing. Restarting hangfire/project/server did not help.

After looking into the code, I can see that the way "LastExecution", "NextExecution", "CreatedAt" time stamps are used in the hash table, winding forward the clock will trigger recurrent jobs premature - which is somewhat OK. But winding back the clock will stall them.

A possible solution could be to detect time stamps in the future in the `RecurringJobScheduler.GetLastInstant`
